Add an early return to updateLayout if unnecessary

colelao opened the pull request:
#### d36800d6828a7eea27126bd52932152b12891888
<pre>
Add an early return to updateLayout if unnecessary
<a href="https://bugs.webkit.org/show_bug.cgi?id=302312">https://bugs.webkit.org/show_bug.cgi?id=302312</a>
<a href="https://rdar.apple.com/164455646">rdar://164455646</a>

Reviewed by Simon Fraser.

There is a recursion updateLayoutIgnorePendingStylesheet -&gt; updateLayout -&gt;
updateRelevancyOfContentVisibilityElements -&gt; updateLayoutIgnorePendingStylesheets.
This recursion makes it possible to handle contentVisibility.

The problem is that updateRelevancyOfContentVisibilityElements calls
updateLayoutIgnorePendingStylesheets as a side effect. When updateLayout tries to
update the content visibility, the expensive updateLayout function is called again.

I have instead given updateLayoutOfContentVisibilityElement&apos;s client&apos;s a choice to
call updateLayoutIgnorePendingStylesheets. That way, we can stop the updateLayout
recursion if updateLayout needs to be called again after
updateLayoutOfContextVisibilityElements.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
(WebCore::Document::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/304566@main">https://commits.webkit.org/304566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98a053201f4a390ae131daeb81f091c465587ed7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135911 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143622 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87470 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dd8f1de0-d80f-4a8b-8633-09d9499a072b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137780 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8941 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8126 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103867 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a7eff8f1-9315-4a82-8335-b198e9c212d3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138857 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121808 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84744 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/74d804b0-28f6-4cdb-971c-99f1f9a5fc3f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6173 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3816 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4226 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115438 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146371 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7965 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40562 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112224 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6687 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112613 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28585 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6087 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118106 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61863 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8012 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36187 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7741 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71564 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7961 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7823 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -369,7 +369,10 @@ TextIterator::TextIterator(const SimpleRange& range, TextIteratorBehaviors behav
 
-    range.start.protectedDocument()->updateLayoutIgnorePendingStylesheets();
+    Ref document = range.start.document();
+    // Only update if style recalc or layout is needed
```
I don't think this comment is useful. Remove?

colelao commented on failures.txt:
```diff
@@ -0,0 +1,5 @@
+editing/execCommand/insert-line-break-no-scroll.html
```
I will remove these.

Andrew-Fryer replied:
@colelao did you mean to commit this?

vitorroriz commented on failures.txt:
oops, I guess you accidentally pushed this?

colelao replied:
Yes, whoops!

Andrew-Fryer commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3064,11 +3065,36 @@ bool Document::updateStyleIfNeededIgnoringPendingStylesheets()
 auto Document::updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> layoutOptions, const Element* context) -> UpdateLayoutResult
 {
-    layoutOptions.add(LayoutOptions::IgnorePendingStylesheets);
+    using enum LayoutOptions;
```
I'm not completely sure this obeys the style guide, but I think it does and IMHO, it makes sense here.
https://webkit.org/code-style-guidelines/

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -235,7 +235,8 @@ void ContentVisibilityDocumentState::updateContentRelevancyForScrollIfNeeded(con
         // Since we may not have determined initial visibility yet, force scheduling the content relevancy update.
         scrollAnchorRoot->protectedDocument()->scheduleContentRelevancyUpdate(ContentRelevancy::OnScreen);
-        scrollAnchorRoot->protectedDocument()->updateRelevancyOfContentVisibilityElements();
+        if (scrollAnchorRoot->protectedDocument()->updateRelevancyOfContentVisibilityElements() == DidUpdateAnyContentRelevancy::Yes)
```
Put `scrollAnchorRoot->protectedDocument()` into a local RefPtr.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3062,7 +3062,8 @@ bool Document::updateStyleIfNeededIgnoringPendingStylesheets()
         scheduleFullStyleRebuild();
 
-    updateRelevancyOfContentVisibilityElements();
+    if (updateRelevancyOfContentVisibilityElements() == DidUpdateAnyContentRelevancy::Yes)
```
All the callers (but one) of `updateRelevancyOfContentVisibilityElements()` now need to do this dance of checking the return value and calling `updateLayoutIgnorePendingStylesheets()`. So I think you should move that back inside `updateLayoutIgnorePendingStylesheets()`, but then do something so that the one special call site, in `Document::updateLayout`, can do the optimal thing.

colelao committed:
Add an early return to updateLayout if unnecessary
https://bugs.webkit.org/show_bug.cgi?id=302312
rdar://164455646

Reviewed by Simon Fraser.

There is a recursion updateLayoutIgnorePendingStylesheet -> updateLayout ->
updateRelevancyOfContentVisibilityElements -> updateLayoutIgnorePendingStylesheets.
This recursion makes it possible to handle contentVisibility.

The problem is that updateRelevancyOfContentVisibilityElements calls
updateLayoutIgnorePendingStylesheets as a side effect. When updateLayout tries to
update the content visibility, the expensive updateLayout function is called again.

I have instead given updateLayoutOfContentVisibilityElement's client's a choice to
call updateLayoutIgnorePendingStylesheets. That way, we can stop the updateLayout
recursion if updateLayout needs to be called again after
updateLayoutOfContextVisibilityElements.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout):
(WebCore::Document::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/304566@main
